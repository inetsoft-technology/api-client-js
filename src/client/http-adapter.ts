/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Interface for classes that provide an adapter to an HTTP client implementation.
 */
export interface HttpAdapter<R> {
   /**
    * Executes an HTTP GET request.
    *
    * @param url     the URL.
    * @param params  the request query parameters, if any.
    * @param headers the HTTP request headers, if any.
    *
    * @return the HTTP response body.
    */
   get(url: string, params?: {[name: string]: string},
       headers?: {[name: string]: string}): Promise<R>;

   /**
    * Executes an HTTP POST request.
    *
    * @param url     the URL.
    * @param body    the HTTP request body.
    * @param params  the request query parameters, if any.
    * @param headers the HTTP request headers, if any.
    *
    * @return the HTTP response body.
    */
   post(url: string, body?: any, params?: {[name: string]: string},
        headers?: {[name: string]: string}): Promise<R>;

   /**
    * Executes an HTTP PUT request.
    *
    * @param url     the URL.
    * @param body    the HTTP request body.
    * @param params  the request query parameters, if any.
    * @param headers the HTTP request headers, if any.
    *
    * @return the HTTP response body.
    */
   put(url: string, body?: any, params?: {[name: string]: string},
       headers?: {[name: string]: string}): Promise<R>;

   /**
    * Executes an HTTP GET request.
    *
    * @param url     the URL.
    * @param headers the HTTP req
    * @param params  the request query parameters, if any.uest headers, if any.
    *
    * @return the HTTP response body.
    */
   delete(url: string, params?: {[name: string]: string},
          headers?: {[name: string]: string}): Promise<R>;

   /**
    * Maps an HTTP response body onto a JSON type.
    *
    * @param response the HTTP response object.
    *
    * @return the HTTP response body.
    */
   map<T>(response: Promise<R>): Promise<T>;

   /**
    * Maps an HTTP response body onto plain text.
    *
    * @param response the HTTP response object.
    *
    * @return the HTTP response body.
    */
   mapText(response: Promise<R>): Promise<string>;

   /**
    * Maps an HTTP response body onto raw, binary data.
    *
    * @param response the HTTP response object.
    *
    * @return the HTTP response body.
    */
   mapBinary(response: Promise<R>): Promise<ArrayBuffer>;
}
