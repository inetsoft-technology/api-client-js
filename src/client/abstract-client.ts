/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { HttpAdapter } from "./http-adapter";

/**
 * @hidden
 */
export abstract class AbstractClient {

   private get headers(): Promise<{[name: string]: string}> {
      return this.token().then((t) => ({ "X-Inetsoft-Api-Token": t }));
   }

   protected constructor(private adapter: HttpAdapter<any>, private token: () => Promise<string>,
                         private baseUrl: string) {}

   protected doGet<T>(url: string, params?: { [p: string]: string }): Promise<T> {
      return this.headers.then((h) =>
         this.adapter.map<T>(this.adapter.get(this.getUrl(url), params, h)));
   }

   protected doGetBinary(url: string, params?: { [p: string]: string }): Promise<ArrayBuffer> {
      return this.headers.then((h) =>
         this.adapter.mapBinary(this.adapter.get(this.getUrl(url), params, h)));
   }

   protected doPost<T>(url: string, data?: any, params?: { [p: string]: string }): Promise<T> {
      return this.headers.then((h) =>
         this.adapter.map<T>(this.adapter.post(this.getUrl(url), data, params, h)));
   }

   protected doPostBinary(url: string, data?: any,
                          params?: { [p: string]: string }): Promise<ArrayBuffer>
   {
      return this.headers.then((h) =>
         this.adapter.mapBinary(this.adapter.post(this.getUrl(url), data, params, h)));
   }

   protected doPut<T>(url: string, data?: any, params?: { [p: string]: string }): Promise<T> {
      return this.headers.then((h) =>
         this.adapter.map<T>(this.adapter.put(this.getUrl(url), data, params, h)));
   }

   protected doDelete<T>(url: string, params?: { [p: string]: string }): Promise<T> {
      return this.headers.then((h) =>
         this.adapter.map<T>(this.adapter.delete(this.getUrl(url), params, h)));
   }

   private getUrl(relativeUrl: string): string {
      return `${this.baseUrl}${relativeUrl}`;
   }
}
