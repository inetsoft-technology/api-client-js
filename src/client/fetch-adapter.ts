/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { HttpAdapter } from "./http-adapter";

/**
 * @hidden
 */
export class FetchAdapter implements HttpAdapter<Response> {
   public get(url: string, params?: { [p: string]: string },
              headers?: { [p: string]: string }): Promise<Response>
   {
      const init = {method: "GET", headers};
      return fetch(this.appendParameters(url, params), init);
   }

   public post(url: string, body?: any, params?: { [p: string]: string },
               headers?: { [p: string]: string }): Promise<Response>
   {
      const init = {
         body: JSON.stringify(body),
         headers,
         method: "POST"
      };
      return fetch(this.appendParameters(url, params), init);
   }

   public put(url: string, body?: any, params?: { [p: string]: string },
              headers?: { [p: string]: string }): Promise<Response>
   {
      const init = {
         body: JSON.stringify(body),
         headers,
         method: "PUT"
      };
      return fetch(this.appendParameters(url, params), init);
   }

   public delete(url: string, params?: { [p: string]: string },
                 headers?: { [p: string]: string }): Promise<Response>
   {
      const init = {method: "DELETE", headers};
      return fetch(this.appendParameters(url, params), init);
   }

   public map<T>(response: Promise<Response>): Promise<T> {
      return response.then((r) => r.json());
   }

   public mapText(response: Promise<Response>): Promise<string> {
      return response.then((r) => r.text());
   }

   public mapBinary(response: Promise<Response>): Promise<ArrayBuffer> {
      return response.then((r) => r.arrayBuffer());
   }

   private appendParameters(url: string, params?: { [p: string]: string }): string {
      if (params) {
         const keys = Object.keys(params);

         if (keys.length) {
            let query = "?";

            for (let i = 0; i < keys.length; i++) {
               if (i > 0) {
                  query += "&";
               }

               query += encodeURIComponent(keys[i]) + "=" + encodeURIComponent(params[keys[i]]);
            }

            return url + query;
         }
      }

      return url;
   }
}
