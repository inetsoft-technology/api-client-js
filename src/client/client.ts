/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { DatabaseApi } from "./database/database-api";
import { DatabaseClient } from "./database/database-client";
import { DataSourceApi } from "./datasource/data-source-api";
import { DataSourceClient } from "./datasource/data-source-client";
import { FetchAdapter } from "./fetch-adapter";
import { FileApi } from "./file/file-api";
import { FileClient } from "./file/file-client";
import { HttpAdapter } from "./http-adapter";
import { PrinterApi } from "./printer/printer-api";
import { PrinterClient } from "./printer/printer-client";
import { ReportApi } from "./report/report-api";
import { ReportClient } from "./report/report-client";
import { ScheduleApi } from "./schedule/schedule-api";
import { ScheduleClient } from "./schedule/schedule-client";
import { ServerApi } from "./server/server-api";
import { ServerClient } from "./server/server-client";
import { ViewsheetApi } from "./sheet/viewsheet/viewsheet-api";
import { ViewsheetClient } from "./sheet/viewsheet/viewsheet-client";
import { WorksheetApi } from "./sheet/worksheet/worksheet-api";
import { WorksheetClient } from "./sheet/worksheet/worksheet-client";

/**
 * @hidden
 */
class TokenProvider {
   private token: string | undefined;

   constructor(private serverUrl: string, private username: string | null,
               private password: string | null, private adapter: HttpAdapter<any>)
   {
   }

   public get(): Promise<string> {
      if(this.token) {
         return Promise.resolve(this.token);
      }

      const url = `${this.serverUrl}api/public/login`;
      let response: Promise<any> | undefined;

      if(this.username) {
         const data = {
            username: this.username,
            password: this.password
         };
         response = this.adapter.post(url, data);
      }
      else {
         response = this.adapter.get(url);
      }

      return this.adapter.map<{token: string}>(response).then(data => {
         this.token = data.token;
         return data.token;
      });
   }
}

/**
 * ApiClient is the client for the InetSoft Web APIs.
 */
export interface ApiClient {
   /**
    * The database API client.
    */
   database: DatabaseApi;

   /**
    * The data source API client.
    */
   dataSource: DataSourceApi;

   /**
    * The file API client.
    */
   file: FileApi;

   /**
    * The printer API client.
    */
   printer: PrinterApi;

   /**
    * The report API client.
    */
   report: ReportApi;

   /**
    * The schedule API client.
    */
   schedule: ScheduleApi;

   /**
    * The server API client.
    */
   server: ServerApi;

   /**
    * The viewsheet API client.
    */
   viewsheet: ViewsheetApi;

   /**
    * The worksheet API client.
    */
   worksheet: WorksheetApi;
}

/**
 * Implementation of the ApiClient interface.
 *
 * @hidden
 */
class ApiClientImpl implements ApiClient {

   public readonly database: DatabaseApi;
   public readonly dataSource: DataSourceApi;
   public readonly file: FileApi;
   public readonly printer: PrinterApi;
   public readonly report: ReportApi;
   public readonly schedule: ScheduleApi;
   public readonly server: ServerApi;
   public readonly viewsheet: ViewsheetApi;
   public readonly worksheet: WorksheetApi;

   constructor(serverUrl: string, username: string | null, password: string | null,
               adapter: HttpAdapter<any> | null)
   {
      adapter = adapter || new FetchAdapter();
      const provider = new TokenProvider(serverUrl, username, password, adapter);
      const token = () => provider.get();
      this.database = new DatabaseClient(adapter, token, serverUrl);
      this.dataSource = new DataSourceClient(adapter, token, serverUrl);
      this.file = new FileClient(adapter, token, serverUrl);
      this.printer = new PrinterClient(adapter, token, serverUrl);
      this.report = new ReportClient(adapter, token, serverUrl);
      this.schedule = new ScheduleClient(adapter, token, serverUrl);
      this.server = new ServerClient(adapter, token, serverUrl);
      this.viewsheet = new ViewsheetClient(adapter, token, serverUrl);
      this.worksheet = new WorksheetClient(adapter, token, serverUrl);
   }
}

/**
 * Creates a new instance of ApiClient.
 *
 * @param serverUrl the base URL of the InetSoft server.
 * @param username  the user name used to authenticate with the server. This parameter is not
 *                  required if anonymous access is allowed or if the user is already
 *                  authenticated in the enclosing web page.
 * @param password  the password used to authenticate with the server.
 * @param adapter   the adapter for the HTTP client. If not specified, this will use the fetch API.
 *
 * @return an new API client instance.
 */
export function createApiClient(serverUrl: string, username: string | null = null,
                                password: string | null = null,
                                adapter: HttpAdapter<any> | null = null): ApiClient {
   if(!serverUrl || !serverUrl.length) {
      throw new Error("The serverUrl parameter is required");
   }

   if(serverUrl[serverUrl.length - 1] !== "/") {
      serverUrl = serverUrl + "/";
   }

   return new ApiClientImpl(serverUrl, username, password, adapter);
}
