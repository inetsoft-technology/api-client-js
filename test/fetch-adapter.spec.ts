/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FetchAdapter } from "../src/client/fetch-adapter";

declare const global: any;

describe("FetchAdapter", () => {
   let adapter: FetchAdapter;

   beforeEach(() => {
      global.fetch = jest.fn();
      adapter = new FetchAdapter();
   });

   describe("when get() is called", () => {
      it("fetch is called with the correct arguments", () => {
         global.fetch.mockReturnValueOnce(Promise.resolve({status: 200}));
         return adapter.get("http://localhost:8080/sree/test", {p1: "v1", p2: "v2"}, {"X-Inetsoft-Token": "secret"}).then(response => {
            expect(response).toMatchSnapshot();
            expect(global.fetch.mock.calls.length).toBe(1);
            expect(global.fetch.mock.calls[0][0]).toBe("http://localhost:8080/sree/test?p1=v1&p2=v2");
            expect(global.fetch.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when get() is called with no parameters", () => {
      it("fetch is called with the correct arguments", () => {
         global.fetch.mockReturnValueOnce(Promise.resolve({status: 200}));
         return adapter.get("http://localhost:8080/sree/test", undefined, {"X-Inetsoft-Token": "secret"}).then(response => {
            expect(response).toMatchSnapshot();
            expect(global.fetch.mock.calls.length).toBe(1);
            expect(global.fetch.mock.calls[0][0]).toBe("http://localhost:8080/sree/test");
            expect(global.fetch.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when get() is called with empty parameters", () => {
      it("fetch is called with the correct arguments", () => {
         global.fetch.mockReturnValueOnce(Promise.resolve({status: 200}));
         return adapter.get("http://localhost:8080/sree/test", {}, {"X-Inetsoft-Token": "secret"}).then(response => {
            expect(response).toMatchSnapshot();
            expect(global.fetch.mock.calls.length).toBe(1);
            expect(global.fetch.mock.calls[0][0]).toBe("http://localhost:8080/sree/test");
            expect(global.fetch.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when post() is called", () => {
      it("fetch is called with the correct arguments", () => {
         global.fetch.mockReturnValueOnce(Promise.resolve({status: 200}));
         return adapter.post("http://localhost:8080/sree/test", {query: "value"}, {p1: "v1", p2: "v2"}, {"X-Inetsoft-Token": "secret"}).then(response => {
            expect(response).toMatchSnapshot();
            expect(global.fetch.mock.calls.length).toBe(1);
            expect(global.fetch.mock.calls[0][0]).toBe("http://localhost:8080/sree/test?p1=v1&p2=v2");
            expect(global.fetch.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when put() is called", () => {
      it("fetch is called with the correct arguments", () => {
         global.fetch.mockReturnValueOnce(Promise.resolve({status: 200}));
         return adapter.put("http://localhost:8080/sree/test", {query: "value"}, {p1: "v1", p2: "v2"}, {"X-Inetsoft-Token": "secret"}).then(response => {
            expect(response).toMatchSnapshot();
            expect(global.fetch.mock.calls.length).toBe(1);
            expect(global.fetch.mock.calls[0][0]).toBe("http://localhost:8080/sree/test?p1=v1&p2=v2");
            expect(global.fetch.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when delete() is called", () => {
      it("fetch is called with the correct arguments", () => {
         global.fetch.mockReturnValueOnce(Promise.resolve({status: 200}));
         return adapter.delete("http://localhost:8080/sree/test", {p1: "v1", p2: "v2"}, {"X-Inetsoft-Token": "secret"}).then(response => {
            expect(response).toMatchSnapshot();
            expect(global.fetch.mock.calls.length).toBe(1);
            expect(global.fetch.mock.calls[0][0]).toBe("http://localhost:8080/sree/test?p1=v1&p2=v2");
            expect(global.fetch.mock.calls[0][1]).toMatchSnapshot();
         });
      });
   });

   describe("when map() is called", () => {
      it("the promise is resolved to the response JSON body", () => {
         const response: any = {
            json: jest.fn(() => ({result: "value"}))
         };
         return adapter.map(Promise.resolve(response)).then(result => {
            expect(result).toMatchSnapshot();
            expect(response.json.mock.calls.length).toBe(1);
         });
      });
   });

   describe("when mapText() is called", () => {
      it("the promise is resolved to the response text body", () => {
         const response: any = {
            text: jest.fn(() => "value")
         };
         return adapter.mapText(Promise.resolve(response)).then(result => {
            expect(result).toEqual("value");
            expect(response.text.mock.calls.length).toBe(1);
         });
      });
   });

   describe("when mapBinary() is called", () => {
      it("the promise is resolved to the response binary body", () => {
         const response: any = {
            arrayBuffer: jest.fn(() => createBinaryData())
         };
         return adapter.mapBinary(Promise.resolve(response)).then(result => {
            validateBinaryData(result);
            expect(response.arrayBuffer.mock.calls.length).toBe(1);
         });
      });
   });
});

function createBinaryData(): ArrayBuffer {
   const buffer = new ArrayBuffer(16);
   const array = new Int32Array(buffer);
   array[0] = 1;
   array[1] = 2;
   array[2] = 3;
   array[3] = 4;
   return buffer;
}

function validateBinaryData(buffer: ArrayBuffer): void {
   expect(buffer).toBeTruthy();
   expect(buffer.byteLength).toBe(16);
   const actual = new Int32Array(buffer);
   expect(actual[0]).toBe(1);
   expect(actual[1]).toBe(2);
   expect(actual[2]).toBe(3);
   expect(actual[3]).toBe(4);
}
