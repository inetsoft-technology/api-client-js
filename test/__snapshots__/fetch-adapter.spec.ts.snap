// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchAdapter when delete() is called fetch is called with the correct arguments 1`] = `
Object {
  "status": 200,
}
`;

exports[`FetchAdapter when delete() is called fetch is called with the correct arguments 2`] = `
Object {
  "headers": Object {
    "X-Inetsoft-Token": "secret",
  },
  "method": "DELETE",
}
`;

exports[`FetchAdapter when get() is called fetch is called with the correct arguments 1`] = `
Object {
  "status": 200,
}
`;

exports[`FetchAdapter when get() is called fetch is called with the correct arguments 2`] = `
Object {
  "headers": Object {
    "X-Inetsoft-Token": "secret",
  },
  "method": "GET",
}
`;

exports[`FetchAdapter when get() is called with empty parameters fetch is called with the correct arguments 1`] = `
Object {
  "status": 200,
}
`;

exports[`FetchAdapter when get() is called with empty parameters fetch is called with the correct arguments 2`] = `
Object {
  "headers": Object {
    "X-Inetsoft-Token": "secret",
  },
  "method": "GET",
}
`;

exports[`FetchAdapter when get() is called with no parameters fetch is called with the correct arguments 1`] = `
Object {
  "status": 200,
}
`;

exports[`FetchAdapter when get() is called with no parameters fetch is called with the correct arguments 2`] = `
Object {
  "headers": Object {
    "X-Inetsoft-Token": "secret",
  },
  "method": "GET",
}
`;

exports[`FetchAdapter when map() is called the promise is resolved to the response JSON body 1`] = `
Object {
  "result": "value",
}
`;

exports[`FetchAdapter when post() is called fetch is called with the correct arguments 1`] = `
Object {
  "status": 200,
}
`;

exports[`FetchAdapter when post() is called fetch is called with the correct arguments 2`] = `
Object {
  "body": "{\\"query\\":\\"value\\"}",
  "headers": Object {
    "X-Inetsoft-Token": "secret",
  },
  "method": "POST",
}
`;

exports[`FetchAdapter when put() is called fetch is called with the correct arguments 1`] = `
Object {
  "status": 200,
}
`;

exports[`FetchAdapter when put() is called fetch is called with the correct arguments 2`] = `
Object {
  "body": "{\\"query\\":\\"value\\"}",
  "headers": Object {
    "X-Inetsoft-Token": "secret",
  },
  "method": "PUT",
}
`;
